safe/gis/tools.py

```diff
@@ -113,45 +129,194 @@ def reclassify_value(one_value, ranges):
+                    -1]  # If path was absolute
+                extension = os.path.splitext(file_name)[1]
+                if extension in OGR_EXTENSIONS:
+                    provider = 'ogr'
+                elif extension in GDAL_EXTENSIONS:
+                    provider = 'gdal'
+                else:
+                    provider = None
+
+    if not provider:
+        layer = load_layer_without_provider(layer_path)
+    else:
+        layer = load_layer_with_provider(layer_path, provider)
+
+    if not layer or not layer.isValid():
+        message = 'Layer "%s" is not valid' % layer_path
+        LOGGER.debug(message)
+        raise Exception(message)
+
+    # Define the name
+    if not name:
+        source = layer.source()
+        if '|' in source:
+            clean_uri = source.split('|')[0]
+        else:
+            clean_uri = source
+        is_file_based = os.path.exists(clean_uri)
+        if is_file_based:
+            # Extract basename and absolute path
+            file_name = os.path.split(layer_path)[-1]  # If path was absolute
+            name = os.path.splitext(file_name)[0]
+        else:
+            # Might be a DB, take the DB name
+            source = QgsDataSourceURI(source)
+            name = source.table()
+
+    if not name:
+        name = 'default'
+
+    if qgis_version() >= 21800:
+        layer.setName(name)
+    else:
+        layer.setLayerName(name)
 
     # Determine if layer is hazard or exposure
     layer_purpose = 'undefined'
     try:
-        keywords = read_iso19115_metadata(layer_path)
+        keywords = read_iso19115_metadata(layer.source())
         if 'layer_purpose' in keywords:
             layer_purpose = keywords['layer_purpose']
     except NoKeywordsFoundError:
         pass
 
-    # Create QGis Layer Instance
-    if extension in ['.asc', '.tif', '.tiff']:
-        layer = QgsRasterLayer(layer_path, base_name)
-    elif extension in ['.shp', '.geojson', '.gpkg']:
-        layer = QgsVectorLayer(layer_path, base_name, 'ogr')
-    elif extension in ['.csv']:
-        uri = 'file:///%s?delimiter=%s' % (layer_path, ',')
-        layer = QgsVectorLayer(uri, base_name, 'delimitedtext')
+    monkey_patch_keywords(layer)
+
+    return layer, layer_purpose
+
+
+def load_layer_with_provider(layer_uri, provider, layer_name='tmp'):
+    """Load a layer with a specific driver.
+
+    :param layer_uri: Layer URI that will be used by QGIS to load the layer.
+    :type layer_uri: basestring
+
+    :param provider: Provider name to use.
+    :type provider: basestring
+
+    :param layer_name: Layer name to use. Default to 'tmp'.
+    :type layer_name: basestring
+
+    :return: The layer or None if it's failed.
+    :rtype: QgsMapLayer
+    """
+    if provider in RASTER_DRIVERS:
+        return QgsRasterLayer(layer_uri, layer_name, provider)
+    elif provider in VECTOR_DRIVERS:
+        return QgsVectorLayer(layer_uri, layer_name, provider)
```

Comment:
What about delimitedtext? I saw in the other part of code, we need to use it explicitly.
```
            # Explicitly use URI with delimiter or tests fail in Windows. TS.
            layer = QgsVectorLayer(
                'file:///%s?delimiter=,' % layer_uri, layer_name, driver)
```

Reply:
Yes, I was thinking about this one. But this function is used when the caller of this function known the provider name. So if the user known the provider name, I expected the caller to have a well formatted  URI

Reply:
ah I see. But perhaps put the expected URI for this special  case (delimited text) to make the future usage of this method easier.